types/time.go

```diff
@@ -1753,41 +1753,82 @@ func extractDayMinute(format string) (int64, int64, int64, float64, error) {
 	return 0, 0, days, value, nil
 }
 
-// extractDayHour extracts day and hour from a string and its format is `DD HH`.
+// extractDayHour extracts day and hour from a string and its suggested format is `DD HH`.
+// MySQL permits any punctuation delimiter in the expr format. it may start with a - for negative intervals
 func extractDayHour(format string) (int64, int64, int64, float64, error) {
-	fields := strings.Split(format, " ")
-	if len(fields) != 2 {
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
+	if len(format) == 0 {
+		return 0, 0, 0, 0, nil
 	}
+	neg := false
+	format0 := format
+	if format[0] == '-' {
+		neg = true
+		format = format[1:]
+	}
+	fields := numericRegex.FindAllString(format, -1)
 
-	days, err := strconv.ParseInt(fields[0], 10, 64)
-	if err != nil {
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
+	if len(fields) == 0 {
+		return 0, 0, 0, 0, nil
+	} else if len(fields) > 2 {
+		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format0)
+	}
+	field0 := "0"
+	if len(fields) > 1 {
+		field0 = fields[len(fields)-2]
 	}
+	field1 := fields[len(fields)-1]
 
-	hours, err := strconv.ParseFloat(fields[1], 64)
+	days, err := strconv.ParseInt(field0, 10, 64)
 	if err != nil {
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
+		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format0)
+	}
+	hours, err := strconv.ParseFloat(field1, 64)
+	if err != nil {
+		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format0)
+	}
+	if neg {
+		days *= -1
+		hours *= -1
 	}
-
 	return 0, 0, days, hours * float64(gotime.Hour), nil
 }
 
-// extractYearMonth extracts year and month from a string and its format is `YYYY-MM`.
+// extractYearMonth extracts year and month from a string and its suggested format is `YYYY-MM`.
+// MySQL permits any punctuation delimiter in the expr format. it may start with a - for negative intervals
 func extractYearMonth(format string) (int64, int64, int64, float64, error) {
```

Comment:
looks like the main logic of `extractDayHour` and `extractYearMonth` is the same? Can we extract a function and reuse these codes?

Reply:
@zz-jason 
I use extract a function to handle all time format.
PTAL